updated milestones document (#260)
open-telemetry/opentelemetry-specification: milestones.md

a few major areas of improvement - change for out of band span reporting API and
the merged pre-aggregated and non-aggregated metrics API. There are other
changes as well.

We already plan for the next iteraion and we know that API is expected to be
changed for the Alpha v0.3 milestone. The big improvements in works are
initialization and configuring logic as well as context propagation detachment.

#### SDK specs for Alpha v0.2

There were no formal SDK proposal. So this milestone will define SDK data
structures and public methods for the first time.

In scope of SDK Alpha release are:

- Basic telemetry pipeline for traces complete
  - Built-in samplers (percentage sampler).
  - SpanProcessor interface and implementations:
  - Default and built-in processors
    - Batching processors with the options to block when the queue is full
      processor
  - Exporter interface
- Distributed context
  - Basic implementation
- Metrics
  - Metrics processor
  - Metrics exporting interface
  - Default and built-in processors
    - Aggregation processor
  - Metrics aggregation implementation

Also in scope:

- Jaeger and/or Zipkin exporter
- Prometheus exporter

> - Proto definition and optionally - OpenTelemetry collector exporter

Reviewer: >OpenTelemetry collector exporter

Can you please clarify this one? Does this mean an exporter implementation in the SDK that is capable of sending data to OpenTelemetry Collector? Note that Collector can accept multiple protocols, including Jaeger, Zipkin, OpenCensus, Prometheus so technically if we have those exporters then we can export to OpenTelemetry Collector. However we do not yet have a new OpenTelemetry protocol defined (my proposal is under review: https://github.com/open-telemetry/oteps/pull/35), so I am guessing this is what this refers to. Please confirm.

Author: yes, I added it to the next milestone and haven't removed it here. Sorry for confusion

### Alpha v0.2 release validation

Languages will ship alpha releases shortly after specification will be complete.
The main purpose of this release is to start implementing data collectors and
use API and SDK for the user scenarios. This will ensure validation of concepts
and public surface.

Note, we DO expect changes in APIs for the Alpha v0.3 release.

### Alpha v0.3 release

The spirit of beta release is to deliver a product with the stable, almost the
release candidate level of APIs and interfaces.

After the beta release of language SDKs we do expect that languages public
surface may change, but we do not expect any major changes in conceptual level
in specifications.

**Porposed deadline**: Specification complete by Nov 15th.

> In scope of SDK Alpha v0.2 release are:

Reviewer 2: ```suggestion
In scope of SDK Alpha v0.3 release are:
```

> -	Collector support for Otel protocol

Reviewer: Just to clarify, is this referring to the new OpenTelemetry protocol?

Author: yes

-	Otel exporters

> Required Spec RFCs for Alpha v0.2:

Reviewer 2: ```suggestion
Required Spec RFCs for Alpha v0.3:
```

As OpenTracing and OpenCensus projects converge we write specifications the same
time as we develop libraries.

We will refer to [API
Proposal](https://github.com/open-telemetry/opentelemetry-specification/milestone/1)
Alpha v0.1 release.

### Alpha v0.2

The spirit of this release is to have a demo-able product that anybody can start
playing with and start implementing instrumentation adaptors and data collectors
for various scenarios.

As part of this release we ask to implement Tracing and Metrics exporters so it
will be easy to visualize the scenarios. For Tracing - either Jaeger or Zipkin
can be selected, for Metrics - Prometheus.

> **Proposed deadline:** Alpha v0.2 specs complete by 10/4

Reviewer 3: Additional due date: Alpha v0.2 implementations complete by 10/14

Author: I understand the desire. I would suggest this will be tracked by language SIGs

#### API specs for Alpha v0.2

Alpha v0.1 received a lot of feedback. As part of an Alpha release there will be
a few major areas of improvement - change for out of band span reporting API and
the merged pre-aggregated and non-aggregated metrics API. There are other

> changes as well.

Reviewer 3: Wording feels a bit vague. Since this is a roadmap, focusing on a precise checklist of specs and SDK features in each version, rather than a narrative, might provide clearer guidance to SIG members looking to implement the milestones. 

For the v0.2 spec changes, I believe the following list is complete:

```markdown
The alpha v0.1 specification incorporates the following RFCS:
* [Telemetry Without Manual Instrumentation](https://github.com/open-telemetry/oteps/blob/master/text/0001-telemetry-without-manual-instrumentation.md)
* [Remove SpanData](https://github.com/open-telemetry/oteps/blob/master/text/0002-remove-spandata.md)
* [Consolidate pre-aggregated and raw metrics APIs](https://github.com/open-telemetry/oteps/blob/master/text/0003-measure-metric-type.md)
* [Sampling API](https://github.com/open-telemetry/oteps/blob/master/text/0006-sampling.md)
* [Remove support to report out-of-band telemetry](https://github.com/open-telemetry/oteps/blob/master/text/0007-no-out-of-band-reporting.md)
* [Metrics Observers API](https://github.com/open-telemetry/oteps/blob/master/text/0008-metric-observer.md)
* [Metric Handle API](https://github.com/open-telemetry/oteps/blob/master/text/0009-metric-handles.md)
* [Rename "Cumulative" to "Counter"](https://github.com/open-telemetry/oteps/blob/master/text/0010-cumulative-to-counter.md)
* [Named Tracers and Meters](https://github.com/open-telemetry/oteps/blob/master/text/0016-named-tracers.md)
```

Author: Let's keep it vague and just point to the issues. Thank you for making a triage!

We already plan for the next iteraion and we know that API is expected to be
changed for the Alpha v0.3 milestone. The big improvements in works are
initialization and configuring logic as well as context propagation detachment.

#### SDK specs for Alpha v0.2

There were no formal SDK proposal. So this milestone will define SDK data
structures and public methods for the first time.

> In scope of SDK Alpha release are:

Reviewer 3: For clarity, the SDK list could be simplified. Also, please mention the Alpha version.
```markdown
In scope of the Alpha v0.2 SDK release are:
- Tracing
  - Built-in samplers (percentage sampler).
  - SpanProcessor interface
  - Batching SpanProcessor
  - Exporter interface
  - Jaeger and/or Zipkin exporter
- Context Propagation
  - In-process propagation
  - Inject and Extract 
  - DistributedContext
- Metrics
  - MetricsProcessor interface
  - Aggregation MetricsProcessor
  - Exporter interface
  - Prometheus Exporter

```

Author: thanks!

- Basic telemetry pipeline for traces complete
  - Built-in samplers (percentage sampler).

>   - SpanProcessor interface and implementations:

Reviewer 3: There has been a request from the Golang SIG to move the SpanProcessor interface to the Alpha v0.3 deadline. @rghetia can you confirm?

Reviewer 4: We can keep SpanProcessor in v0.2. I am working on the batch processor right now. So it should be done for v0.2 release.

  - Default and built-in processors
    - Batching processors with the options to block when the queue is full
      processor
  - Exporter interface
- Distributed context
  - Basic implementation
- Metrics
  - Metrics processor
  - Metrics exporting interface
  - Default and built-in processors
    - Aggregation processor
  - Metrics aggregation implementation

Also in scope:

- Jaeger and/or Zipkin exporter
- Prometheus exporter

### Alpha v0.2 release validation

Languages will ship alpha releases shortly after specification will be complete.
The main purpose of this release is to start implementing data collectors and
use API and SDK for the user scenarios. This will ensure validation of concepts
and public surface.

Note, we DO expect changes in APIs for the Alpha v0.3 release.

### Alpha v0.3 release

The spirit of beta release is to deliver a product with the stable, almost the
release candidate level of APIs and interfaces.

After the beta release of language SDKs we do expect that languages public
surface may change, but we do not expect any major changes in conceptual level

> in specifications.

Reviewer 3: The description for the  Beta milestone looks like it got mixed up with the Alpha v0.3 milestone.

Author: thanks!

**Proposed deadline**: Specification complete by Nov 15th.

In scope of SDK Alpha v0.3 release are:

- Collector support for OpenTelemetry protocol

> - OpenTelemetry exporters

Reviewer: Depending on how quickly we approve the protocol specification we may not have enough time to also implement it in OpenTelemetry Collector and the SDK Exporters. I would suggest to do this: in v0.3 we complete the protocol specification, in v0.4 we implement SDK Exporters and add the protocol support to Collector.

Also, FYI, we are planning to make OpenTelemetry Collector Core Beta release on Oct 18 (https://github.com/open-telemetry/opentelemetry-service/milestone/2). The criteria for Beta is stability of interfaces and of configuration file format and extensive test coverage for all non-trivial functionality.

The implementation of OpenTelemetry Protocol in my opinion is not a blocker for Collector Beta release. The only concern would be if there is a possibility of confusion if we declare Collector "Beta" and the rest of project still uses "Alpha" label. If this sounds like an issue we can keep the same labeling for Collector and will call it "Alpha v0.x" like the rest of the project does.

Reviewer 3: Thanks @tigrannajaryan . I think that is a reasonable suggestion, it will allow us to ship a `v0.3` faster.

I would suggest we call the Collector `alpha` for now, until we hit `v0.4`, to keep the messaging simple, but it's not a strong opinion.

    - Batching processors
    - Block when the queue is full processor
  - Exporter interface
- Distributed context
  - Basic implementation
- Metrics
  - Metrics aggregation implementation
  - Metrics exporting interface

Also in scope:

- Jaeger and/or Zipkin exporter
- Prometheus exporter

> - Proto definition and optionally - OpenTelemetry Agent exporter

Reviewer 2: ```suggestion
- Proto definition and optionally - OpenTelemetry Collector exporter
```
See https://github.com/open-telemetry/opentelemetry-service/issues/352.

**Proposed deadline:** SDK specs complete by 10/4

### Alpha release validation

Languages will ship alpha releases shortly after specification will be complete.
The main purpose of this release is to start implementing data collectors and
use API and SDK for the user scenarios. This will ensure validation of concepts
and public surface.

Note, we DO expect changes in APIs for the beta release.

### Beta release

The spirit of beta release is to deliver a product with the stable, almost the
release candidate level of APIs and interfaces.

After the beta release of language SDKs we do expect that languages public
surface may change, but we do not expect any major changes in conceptual level
in specifications.

> **Porposed deadline**: Specification complete by Nov 15th.

Reviewer 5: ```suggestion
**Proposed deadline**: Specification complete by Nov 15th.
```

Reviewer 2: Seems this commit is reverted by mistake. Please re-commit.

**Details of this release to be added later.**

### Getting to release

> By end of year mid August the Java basic SDK will be complete and we will begin

Reviewer 2: ```suggestion
By end of year the basic language SDKs will be complete and we will begin
```

Reviewer 2: Same here.
